packages/lit-html/src/disconnectable-directive.ts

```diff
@@ -0,0 +1,350 @@
+ *                       ._directive
+ *                         DisconnectableDirective(D24)
+ *                   NodePart(N25) _$dC=[D26]
+ *                    ._directive
+ *                      DisconnectableDirective(D26)
+ *                    ._value
+ *                      string
+ *
+ * Example 1: The directive in NodePart(N12) updates and returns `nothing`. The
+ * NodePart will _clear() itself, and so we need to disconnect the "value" of
+ * the NodePart (but not its directive). In this case, when `_clear()` calls
+ * `_$setNodePartConnected()`, we don't iterate all of the
+ * _$disconnectableChildren, rather we do a value-specific disconnection: i.e.
+ * since the _value was an Array<NodePart> (because an iterable had been
+ * committed), we iterate the array of NodeParts (N14, N16) and run
+ * `setConnected` on them (which does recurse down the full tree of
+ * `_$disconnectableChildren` below it, and also removes N14 and N16 from N12's
+ * `_$disconnectableChildren`). Once the values have been disconnected, we then
+ * check whether the NodePart(N12)'s list of `_$disconnectableChildren` is empty
+ * (and would remove it from its parent TemplateInstance(T3) if so), but since
+ * it would still contain its directive D13, it stays in the disconnectable
+ * tree.
+ *
+ * Example 2: In the course of Example 1, `setConnected` will reach
+ * NodePart(N16); in this case the entire part is being disconnected, so we
+ * simply iterate all of N16's `_$disconnectableChildren` (D17,T18) and
+ * recursively run `setConnected` on them. Note that we only remove children
+ * from `_$disconnectableChildren` for the top-level values being disconnected
+ * on a clear; doing this bookkeeping lower in the tree is wasteful since it's
+ * all being thrown away.
+ *
+ * Example 3: If the LitElement containing the entire tree above becomes
+ * disconnected, it will run `setDirectiveConnection()` (which calls
+ * `part._$setNodePartConnected()` if it exists); in this case, we recursively
+ * run `setConnected()` over the entire tree, without removing any children from
+ * `_$disconnectableChildren`, since this tree is required to re-connect the
+ * tree, which does the same operation, simply passing `isConnectd: true` down
+ * the tree, signaling which callback to run.
+ */
+
+import {
+  Directive,
+  PartInfo,
+  NodePart,
+  NODE_PART,
+  DisconnectableParent,
+} from './lit-html.js';
+export {directive} from './lit-html.js';
+
+/**
+ * Recursively walks down the tree of Parts/TemplateInstances/Directives to set
+ * the connected state of directives and run `disconnectedCallback`/
+ * `reconnectedCallback`s.
+ */
+const setChildrenConnected = (
+  parent: DisconnectableParent,
+  isConnected: boolean,
+  shouldRemoveFromParent = false
+) => {
+  const children = parent._$disconnetableChildren;
+  if (children !== undefined) {
+    for (const child of children) {
+      if ((child as Directive)._$setDirectiveConnected) {
```

Comment:
Can this be `_$setConnected`?

Reply:
Added comment, it's treated as a brand and disambiguates it from NodePart's `_$setNodePartConnected`

Reply:
If these are brands, then do they need a trailing `$` and be excluded from minification?

Reply:
The thought was that since these are all known objects (`NodePart`, `TemplateInstance`, or `Directive`), they don't need collision-proof brands.

But we need to determine if minifying user `Directive` subclasses as part of application builds could result in collisions with our own minified symbols or not. If they could, it's an issue not of branding, but of overall collision-safety.

I added this issue to track: https://github.com/Polymer/lit-html/issues/1461